status/discover.go

```diff
@@ -91,66 +86,120 @@ func getPathURL(endpoint string) (path string) {
+	// We had the service so we can check the Path
+	path, err := checkIfQueryBasePath(tracingConfig.Namespace, tracingConfig.Service)
+	if err != nil {
+		log.Debugf("[TRACING] Error checking the query base path")
+	}
+	return
+}
 
-	return tracingConfig.URL
+func discoverURLTracingService() (url string) {
+	// Try to discover the URL . Openshift Client
+	url, err := discoverServiceURL(appstate.JaegerConfig.Namespace, appstate.JaegerConfig.Service)
+	if err != nil {
+		log.Debugf("[TRACING] URL discovery failed: %v", err)
+		return
+	}
+	// Trim the string to format correctly the url with the path
+	url = strings.TrimSuffix(url, "/") + "/" + appstate.JaegerConfig.Path
+	appstate.JaegerEnabled = true
+	appstate.JaegerConfig.URL = url
+	return
 }
 
 func DiscoverJaeger() string {
+	// Kiali has all the configuration
 	if appstate.JaegerEnabled {
 		return appstate.JaegerConfig.URL
 	}
+
+	// Get the configuration
 	tracingConfig := config.Get().ExternalServices.Tracing
-	if tracingConfig.URL != "" && tracingConfig.Service != "" {
+
+	// There is not a service in the configuration we need discover the service
+	if tracingConfig.Service == "" {
+		log.Infof("[TRACING] Kiali Service Discovery")
+		service := discoverTracingService()
+		if service != "" {
+			tracingConfig.Service = service
+			log.Infof("[TRACING] Kiali found tracing service in %s", tracingConfig.Service)
+		} else {
+			log.Infof("[TRACING] Service discovery failed: Kiali didn't found a tracing service")
+		}
+		appstate.JaegerConfig = tracingConfig
+	}
+
+	//There is an endpoint in the configuration discovery QUERY_BASE_PATH by endpoint defined
+	if tracingConfig.URL != "" {
+		log.Infof("[TRACING] Kiali Path Discovery by user endpoint")
 		// User assumes configuration
 		appstate.JaegerEnabled = true
+		// Get Path from the URL (User could set the QUERY_BASE_PATH in the URL)
+		path := getPathURL(tracingConfig.URL)
+		if path != "" {
+			tracingConfig.Path = path
+			log.Infof("[TRACING] Kiali found tracing Path in the URL defined by the user with value %s", tracingConfig.Path)
+		} else {
+			log.Infof("[TRACING]  Kiali didn't found a Path in the URL defined by the user")
+		}
+		// Save configuration in our appstate
+		appstate.JaegerConfig = tracingConfig
+	}
 
-		tracingConfig.Path = getPathURL(tracingConfig.URL)
+	// Discover QUERY_BASE_PATH by deployment
+	if tracingConfig.Service != "" && tracingConfig.Path == "" {
+		log.Infof("[TRACING] Kiali Path Discovery by deployment definition")
+		path := discoverTracingPath()
+		if path != "" {
+			tracingConfig.Path = path
+			log.Infof("[TRACING] Kiali found tracing QUERY_BASE_PATH defined with value %s", tracingConfig.Path)
+		} else {
+			log.Infof("[TRACING] Tracing Path discovery failed: Kiali didn't found a QUERY_BASE_PATH defined")
+		}
+		appstate.JaegerConfig = tracingConfig
+	}
 
+	//There is not an endpoint, go discover for Openshift
+	if tracingConfig.Service != "" {
```

Comment:
shouldn't we do that only if `tracingConfig.URL == ""` ?

Reply:
Yes you're right, I removed the return there and I forgot add the condition thanks !!!